process/ip_database.go

```diff
@@ -53,33 +56,47 @@ func (p procIPDatabase) Close(ctx context.Context) error {
 	return nil
 }
 
-// Batch processes one or more capsules with the processor. Conditions are
-// optionally applied to the data to enable processing.
-func (p procIPDatabase) Batch(ctx context.Context, capsules ...config.Capsule) ([]config.Capsule, error) {
-	return batchApply(ctx, capsules, p, p.Condition)
-}
+// Create a new IP database processor.
+func newProcIPDatabase(cfg config.Config) (p procIPDatabase, err error) {
+	err = config.Decode(cfg.Settings, &p)
+	if err != nil {
+		return procIPDatabase{}, err
+	}
 
-// Apply processes a capsule with the processor.
-func (p procIPDatabase) Apply(ctx context.Context, capsule config.Capsule) (config.Capsule, error) {
-	// only supports JSON, error early if there are no keys
+	p.operator, err = condition.NewOperator(p.Condition)
+	if err != nil {
+		return procIPDatabase{}, err
+	}
+
+	// only supports JSON, fail if there are no keys
 	if p.Key == "" && p.SetKey == "" {
-		return capsule, fmt.Errorf("process: ip_database: key %s set_key %s: %v", p.Key, p.SetKey, errInvalidDataPattern)
+		return procIPDatabase{}, fmt.Errorf("process: ip_database: key %s set_key %s: %v", p.Key, p.SetKey, errInvalidDataPattern)
 	}
 
-	db, err := ipdb.Get(p.Options)
+	p.db, err = ipdb.Get(p.Options)
 	if err != nil {
-		return capsule, fmt.Errorf("process: ip_database: %v", err)
+		return procIPDatabase{}, fmt.Errorf("process: ip_database: %v", err)
 	}
 
-	// lazy load the database
-	if !db.IsEnabled() {
-		if err := db.Open(ctx); err != nil {
-			return capsule, fmt.Errorf("process: ip_database: %v", err)
+	if !p.db.IsEnabled() {
+		if err := p.db.Open(context.Background()); err != nil {
```

Comment:
refactor: another example here where it'd be better to pass context from the source